File: Medium/322-CoinChange.py
'''
    @author
    Prajjwal Maurya
    Leetcode- https://leetcode.com/pmaurya6jun/
    Leetcode Problem 322.Coin Change
    Language Python
    Level Medium
'''

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount+1]*(amount+1)
        dp[0] = 0
        
        for i in range(1, amount+1):
            for coin in coins:
                if coin <= i:
                    dp[i] = min(dp[i], 1 + dp[i-coin])
        return -1 if dp[-1] == amount+1 else dp[-1]

File: Medium/322-CoinChange.cpp
/*
    @author
    Akshat Atray
    Leetcode- https://leetcode.com/akshatatray/
    Leetcode Problem 322.Coin Change
    Language C++
    Level Medium
*/

// SOLUTION 1
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<long> dp(amount+1, INT_MAX);
        dp[0] = 0;
        for (int i=1; i<=amount; ++i)
            for (auto c : coins)
                if (c <= i)
                    dp[i] = min(dp[i], dp[i-c]+1);
        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};

// SOLUTION 2
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<long> dp(amount+1, INT_MAX);
        dp[0] = 0;
        for (int i=0; i<coins.size(); ++i)
            for (int j=coins[i]; j<=amount; ++j)
                dp[j] = min(dp[j], dp[j - coins[i]]+1);
        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};